### PR TITLE
Feature/support load balancer class attribute

### PR DESCRIPTION
**Pull request by RouxAntoine**
### Type of change

- Enhancement / new feature

### Description

Add support for service spec attribute loadBalancerClass documentation https://kubernetes.io/docs/concepts/services-networking/service/#load-balancer-class, spec https://kubernetes.io/docs/reference/kubernetes-api/service-resources/service-v1/#ServiceSpec

The purpose to support this attribute is to globally allow broker services to use a custom controller class. Theses controller class allow multiple loadbalancer ip provider to live together in same kubernetes cluster.
.i.e. metallb and openstack provider.
 
This change have been discussed on CNCF strimzi-dev slack https://cloud-native.slack.com/archives/C018247K8T0/p1670167684500789
This first version use opportunistically the existing `spec.listeners[*].configuration.class` crd's attribute of kind Kafka.
Now the `configuration.class` could be use for listener of type ingress and for listener of type loadbalancer.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [x] Write tests
- [ ] Make sure all tests pass (unitary test ✅, integration ❌ they failed with openssl error i think is related to local problem on my own computer)
```
 java.lang.RuntimeException: openssl status code 1
        at io.strimzi.certs.OpenSslCertManager$OpensslArgs.exec(OpenSslCertManager.java:650)
        at io.strimzi.certs.OpenSslCertManager$OpensslArgs.exec(OpenSslCertManager.java:608)
        at io.strimzi.certs.OpenSslCertManager.generateCsr(OpenSslCertManager.java:435)
        at io.strimzi.operator.cluster.model.Ca.generateSignedCert(Ca.java:365)
        at io.strimzi.operator.cluster.model.Ca.maybeCopyOrGenerateCerts(Ca.java:503)
        at io.strimzi.operator.cluster.model.ClusterCa.generateBrokerCerts(ClusterCa.java:214)
        at io.strimzi.operator.cluster.model.KafkaCluster.generateCertificatesSecret(KafkaCluster.java:1271)
        at io.strimzi.operator.cluster.operator.assembly.KafkaReconciler.lambda$certificateSecret$48(KafkaReconciler.java:728)
        at io.vertx.core.impl.future.Composition.onSuccess(Composition.java:38)
        at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60)
        at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211)
        at io.vertx.core.impl.future.PromiseImpl.tryComplete(PromiseImpl.java:23)
        at io.vertx.core.impl.future.PromiseImpl.onSuccess(PromiseImpl.java:49)
        at io.vertx.core.impl.future.FutureBase.lambda$emitSuccess$0(FutureBase.java:54)
        at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:174)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:167)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:503)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:995)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:833
```
- [x] Update documentation (not sure if enough, i update documentation with annotation `@Description`)
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] ~Reference relevant issue(s) and close them after merging~
- [x] Update CHANGELOG.md
- [ ] ~Supply screenshots for visual changes, such as Grafana dashboards~

Thanks for review 🙂

### CODE REVIEW

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ListenersValidator.java [hunk 1]
```diff
@@ -240,15 +240,15 @@ private static void validateIpFamilies(Set<String> errors, GenericKafkaListener
...
      * @param listener  Listener which needs to be validated
      */
-    private static void validateIngressClass(Set<String> errors, GenericKafkaListener listener) {
-        if (!KafkaListenerType.INGRESS.equals(listener.getType())
-                && listener.getConfiguration().getIngressClass() != null)    {
-            errors.add("listener " + listener.getName() + " cannot configure ingressClass because it is not Ingress based listener");
+    private static void validateControllerClass(Set<String> errors, GenericKafkaListener listener) {
+        if (!(KafkaListenerType.INGRESS.equals(listener.getType()) || KafkaListenerType.LOADBALANCER.equals(listener.getType()))
+                && listener.getConfiguration().getControllerClass() != null)    {
+            errors.add("listener " + listener.getName() + " cannot configure class because it is not an Ingress or loadbalancer based listener");
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ListenersValidator.java:251]:
```suggestion
            errors.add("listener " + listener.getName() + " cannot configure class because it is not an Ingress or LoadBalancer based listener");
```

File: tools/prerequisites-check.sh [hunk 2]
```diff
@@ -18,7 +18,7 @@ check_command_present "${DOCKER_CMD:-docker}"
 check_command_present shellcheck
 
 # After version 3.3.1, yq --version sends the string to STDERR instead of STDOUT
-YQ_VERSION="$(yq --version 2>&1 | ${SED} 's/^.* //g')"
+YQ_VERSION="$(yq --version 2>&1 | ${SED} -r 's/^.* v?//g')"
```
Comment (scholzj) [inline tools/prerequisites-check.sh:21]: Why was this change needed?

File: api/src/main/java/io/strimzi/api/kafka/model/listener/arraylistener/GenericKafkaListenerConfiguration.java [hunk 3]
```diff
@@ -70,17 +70,19 @@ public void setBrokerCertChainAndKey(CertAndKeySecretSource brokerCertChainAndKe
...
     }
 
-    @Description("Configures the `Ingress` class that defines which `Ingress` controller will be used. " +
-            "This field can be used only with `ingress` type listener. " +
-            "If not specified, the default Ingress controller will be used.")
+    @Description("Configures specific class for `Ingress` and `LoadBalancer` that defines which controller will be used. " +
+            "This field can be used only with `ingress` and `loadbalancer` type listener. " +
+            "If not specified, the default controller will be used." +
+            "For Ingress listener the service spec property `ingressClassName` will be set" +
+            "For LoadBalancer listener the service spec property `loadbalancerClass` will be set")
```
Comment (scholzj) [inline api/src/main/java/io/strimzi/api/kafka/model/listener/arraylistener/GenericKafkaListenerConfiguration.java:77]: The leading spaces are missing + some possible improvements.

```suggestion
            "If not specified, the default controller will be used. " +
            "For Ingress listener the property `ingressClassName` will be set in the Ingress resources. " +
            "For LoadBalancer listener the property `loadBalancerClass` will be set in the Service resources.")
```

**Review by scholzj — COMMENTED** [review]
Thanks for the PR. I left some comments - mostly just minor stuff.

File: tools/prerequisites-check.sh (hunk 2 above)
Comment (RouxAntoine) [reply tools/prerequisites-check.sh:21]: i put it in a separated commit 8837187fadd682b28bcd878169fe25d2f198efa6 : 

```
 fix: yq version on mac os brew package is prefixed with v letter

yq version give `yq (https://github.com/mikefarah/yq/) version v4.30.5`
```

i can remote if you prefer. This is not really related to this feature.

File: api/src/main/java/io/strimzi/api/kafka/model/listener/arraylistener/GenericKafkaListenerConfiguration.java [hunk 4]
```diff
@@ -70,17 +70,19 @@ public void setBrokerCertChainAndKey(CertAndKeySecretSource brokerCertChainAndKe
...
     }
 
-    @Description("Configures the `Ingress` class that defines which `Ingress` controller will be used. " +
-            "This field can be used only with `ingress` type listener. " +
-            "If not specified, the default Ingress controller will be used.")
+    @Description("Configures specific class for `Ingress` and `LoadBalancer` that defines which controller will be used. " +
+            "This field can be used only with `ingress` and `loadbalancer` type listener. " +
+            "If not specified, the default controller will be used. " +
+            "For Ingress listener the property `ingressClassName` will be set in the Ingress resources. " +
+            "For LoadBalancer listener the property `loadBalancerClass` will be set in the Service resources.")
```
Comment (PaulRMellor) [inline api/src/main/java/io/strimzi/api/kafka/model/listener/arraylistener/GenericKafkaListenerConfiguration.java:77]:
```suggestion
    @Description("Configures a specific class for `Ingress` and `LoadBalancer` that defines which controller will be used. " +
            "This field can only be used with `ingress` and `loadbalancer` type listeners. " +
            "If not specified, the default controller is used. " +
            "For an `ingress` listener, set the `ingressClassName` property in the `Ingress` resources. " +
            "For a `loadbalancer` listener, set the `loadBalancerClass` property  in the `Service` resources.")
```

**Review by PaulRMellor — APPROVED** [review]
Thanks. Looks good. I made a suggestion for the description, which should be picked up where the description is reused when rebuilding.

**Comment by scholzj** [comment]
As you changed the `api` module, you will need to re-run the build to update the CRDs ;-) ... that is why the build failed:
```
Run the following to add up-to-date resources:
  mvn clean verify -DskipTests -DskipITs \
    && make crd_install \
    && make helm_install \
    && git add packaging/install/ packaging/helm-charts/ documentation/modules/appendix_crds.adoc cluster-operator/src/main/resources/cluster-roles \
    && git commit -s -m 'Update derived resources'
```

File: tools/prerequisites-check.sh (hunk 2 above)
Comment (scholzj) [reply tools/prerequisites-check.sh:21]: Looks like some change in the latest version of `yq`. `4.25.2` did not seems to have the `v`. Anyway, you can keep the change here - it works for me and it seems to work in the CI, so it should be fine.

**Comment by RouxAntoine** [comment]
> As you changed the `api` module, you will need to re-run the build to update the CRDs ;-) ... that is why the build failed:
> 
> ```
> Run the following to add up-to-date resources:
>   mvn clean verify -DskipTests -DskipITs \
>     && make crd_install \
>     && make helm_install \
>     && git add packaging/install/ packaging/helm-charts/ documentation/modules/appendix_crds.adoc cluster-operator/src/main/resources/cluster-roles \
>     && git commit -s -m 'Update derived resources'
> ```

thanks i fixed it 🙂

**Review by ppatierno — APPROVED** [review]
LGTM. Thanks for the PR!

**Review by scholzj — APPROVED** [review]
LGTM. Thanks for the PR. I will trigger the regression tests for it.

**Comment by RouxAntoine** [comment]
> LGTM. Thanks for the PR. I will trigger the regression tests for it.

thanks for your reactivity

**Comment by scholzj** [comment]
Thanks for the PR.
